Split files and dateonly

jods4 opened the pull request:
Hello @mamift 

I've added two features that I need for my projects, would love if you could merge them into a new release (I've tested a local build in the meantime).

### Implement Split Code by Namespace

I have implemented the existing Split Code By Namespace configuration. (1st commit).
It writes one file for every CLR namespace found in the schemas.
When using this configuration, one must set a new attribute `File` on `<Namespace>` to indicate what output filename should be used for every CLR namespace. I added this because (1) XSD filenames do not map 1:1 to namespaces (there might be multiple namespaces in a single xsd, or elements of a single namespace in multiple xsd); and (2) I did not want to derive the filename from the namespace as those can be long and unwieldy, so I wanted some control.

Not that Split Code By Class is still NOT implemented.

> **Warning**
> *Breaking change*: if someone used `XObjectsCodeGen` as a library, a few public methods now have a different return value (they return an enumeration of namespace+writer instead of a single writer).

> **Warning**
> If consumers of this lib have set the Split Code By Namespace option in their config -- even though it does nothing at the moment -- then they will get an error after updating when they try to generate code.
> This error will tell them that `File` configuration is required when splitting by namespace, so they should either:
> - Remove the Split Code By Namespace config -> the tool would generate exactly the same code as today.
> - Add `File` configuration -> the tool will now generate one file per namespace and respect those filename configs.

> **Note**
> When you have multiple independent root schemas referencing the same shared schema, the output is a little wacky when it comes to `LinqToXsdTypeManager`. 
> The file will be overwritten several times and will eventually reference the `LinqToXsdTypeManager` from the last XSD root file that was generated (and contain none of its own).
> This is not super clean but as far as I can tell the typing information is working properly for any XML root, even though multiple Type Managers are involved. Basically they contain redundant information that is the same in every Type Manager.
> I had no issue in my own tests so far. 
> Note that this only matters if you attempt to generate code for multiple schemas that import a shared common namespace. This is something that wasn't really possible today anyway (at least without hand-modifying the generated files).
> If you find yourself in this situation, a solution to end up with clean code is to generate the shared schema on its own, after all other schemas.

### DateOnly and TimeOnly support

In 2nd commit, I added two new configurations: `UseDateOnly` and `UseTimeOnly`.
When they're used, properties of type `xs:date` will be mapped to .net 6.0 `DateOnly` type; and properties of type `xs:time` will be mapped to `TimeOnly`.

> **Note**
> This feature only supports basic properties. Advanced XSD constructs such as Union types won't support DateOnly.
> This is because the current code + System.Xml types are heavily dependent on DateTime on I don't think I can add support for unions without doing massive changes. So I stopped there.

### Moved NullableReferences option
With the introduction of `UseDateOnly` and `UseTimeOnly`, I didn't want to excessively pollute the global configuration element so I put them into `CodeGeneration`. 
Then I felt bad for introducing `NullableReferences` at the root, so I created another `NullableReferences` inside `CodeGeneration`, that has priority over the old one. The old one is still there for backward compatibility but considered obsolete. All "code generation" options should be under `<CodeGeneration>`.

mamift commented:
Can you:

- Increment the version number in Version.props
- Add at least one test each to test the code generation for the `DateOnly` and `TimeOnly` changes
  - This doesn't have to be too involved, even some simple, contrived XSD file specific for this is enough.

jods4 commented:
@mamift do you prefer 3.4 or 4.0 based on the warnings regarding Split Code by Namespace?

mamift commented:
I'm fine with 3.4

jods4 commented:
@mamift before introducing any new test, I have 22 failures that seem, at first glance, unrelated to my changes.
Is is expected or is `master` supposed to be 100% green?

jods4 commented:
@mamift Ok I finalized the PR:

- Bumped version to 3.4 and updated RELEASENOTES.md
- Regen all LinqToXsd.Schemas. No schema actually changed but I had some trouble. 
Many schemas actually don't compile but they weren't committed so I figured it's well-known and removed them.
NHS CDS schema contains the same partial classes multiple time, it's because it has `includes` and when the schemas are generated one-by-one there's duplication. If the folder is generated at once, only one valid file containing everything is generated and this is what I have committed.
- Added a basic test schema for DateOnly and TimeOnly (both elements and attributes). You can check out the generated code.
I also added a test case that uses this schema to roundtrip (write+read) those values.

jods4 commented:
Something strange is happening with Github... my branch jods4:split-files-and-dateonly contains one more commit (the new test case) that doesn't seem to want to appear in this PR for some reason??
I'm gonna close and re-open.
